driver: gpio: intc: npcx: disable WK_EN before configuring gpio interrupt.

MulinChao opened the pull request:
Although the default state of WK_EN is disabled, the users still have
the chance to configure them when WK_EN is enabled if the target gpio
has been configured and enabled before. This CL makes sure WK_EN is
disabled before the driver configures gpio interrupt in case of
unexpected side effects.

Signed-off-by: Mulin Chao <mlchao@nuvoton.com>

pabigot commented on drivers/interrupt_controller/intc_miwu.c:
```diff
@@ -185,63 +185,61 @@ int npcx_miwu_interrupt_configure(const struct npcx_wui *wui,
-		/* Enable wake-up input sources */
-		NPCX_WKINEN(base, wui->group) |= pmask;
+	/* Enable wake-up input sources */
+	NPCX_WKINEN(base, wui->group) |= pmask;
```
Should these changes be conditional on a non-negative `ret`?

AFAICT `NPCX_MIWU_MODE_DISABLED` is no longer part of the logic here; it used to clear the pmask.  Even if this is not an issue because the GPIO call sites already disabled it, I suspect it should be handled here.

MulinChao replied:
I'm sorry about our ambiguous register name. (I try to add more explanations in each PR possibly without npcx datasheet.)
NPCX_WKEN is "Wake-Up Enable  Register" which is used for enabling/disabling to generate a wake-up signal or interrupt request.
NPCX_WKINEN is "Wake-Up Input Enable x Register" which is used for connecting io-pad and MIWU modules. If it is false, the pin can be left unconnected (i.e., floating).
You need to turn both WKEN and WKINEN on for handling a gpio event. Disabling one of them means turns the interrupt off. But I look this change again. The driver should return -EINVAL directly if the configuration setting is wrong. Then the gpio driver will show an error log and return error code directly.

pabigot replied:
OK, I think I understand what this is doing better.

I suspect the commit should be split into two pieces:

* The change to gpio_npcx, which now uses a simpler configuration approach that turns interrupts off before configuring them;
* The change to intc_npcx, which (AFAICT) is no longer responsible for configuring `NPCX_WK` to disable interrupts, because it expects that to have been done before it's called.

I think the current commit message describes the gpio changes, but nothing describes the intc ones.

If I've reconstructed that correctly, please update this.  Otherwise please update it to explain why both files need to be touched in the same commit (e.g. they're tightly coupled and the system won't work without both).

Also (if I've described the change correctly) have you confirmed that all other callers of `npcx_miwu_interrupt_configure()` have disabled interrupts before it's called?  Or maybe the disable should be pushed down into that module?

MulinChao replied:
> OK, I think I understand what this is doing better.
> 
> I suspect the commit should be split into two pieces:
> 
> * The change to gpio_npcx, which now uses a simpler configuration approach that turns interrupts off before configuring them;
> * The change to intc_npcx, which (AFAICT) is no longer responsible for configuring `NPCX_WK` to disable interrupts, because it expects that to have been done before it's called.
> 
> I think the current commit message describes the gpio changes, but nothing describes the intc ones.
> 
> If I've reconstructed that correctly, please update this. Otherwise please update it to explain why both files need to be touched in the same commit (e.g. they're tightly coupled and the system won't work without both).
>

I will split it into two CLs for better explanations. Thanks.
  
> Also (if I've described the change correctly) have you confirmed that all other callers of `npcx_miwu_interrupt_configure()` have disabled interrupts before it's called? Or maybe the disable should be pushed down into that module?

The default value of WK_EN is zero. The other functions call npcx_miwu_interrupt_configure() during driver init function. Only the gpio driver provides an interface to configure interrupt manually. That's why we saw the side-effect when users configure the same pin twice. But it is safer to add a check for WK_EN in miwu driver.

MulinChao replied:
Done.

pabigot reviewed (changes requested):
`drivers/gpio/gpio_npcx.c` looks OK; I prefer the "disable unconditionally, then conditionally enable" idiom this has moved to.

But something looks wrong with `drivers/interrupt_controller/intc_miwu.c`; see comment.

sjg20 reviewed (approved):
no further comments

dcpleung reviewed (commented):
If you need to upload another changes, could you add `npcx` to your commit summary line? This would make it easier to search through git log.

MulinChao commented:
> If you need to upload another changes, could you add `npcx` to your commit summary line? This would make it easier to search through git log.

Thanks for your suggestions. I will add 'npcx:' in the commit summy line of this and the following CLs.

mnkp commented on drivers/interrupt_controller/intc_miwu.c:
```diff
@@ -186,60 +186,55 @@ int npcx_miwu_interrupt_configure(const struct npcx_wui *wui,
 	const uint32_t base = DRV_CONFIG(miwu_devs[wui->table])->base;
 	uint8_t pmask = BIT(wui->bit);
 
-	if (mode == NPCX_MIWU_MODE_DISABLED) {
-		/* Clear MIWU enable bit */
-		NPCX_WKEN(base, wui->group) &= ~pmask;
```
I'm not familiar with Nuvoton Platform and @MulinChao you are officially listed as a collaborator so certainly know it better than me, however `intc_miwu` is used by multiple drivers, e.g. `uart_npcx`, `wdt_npcx` so removing support for `NPCX_MIWU_MODE_DISABLED` only because we modify `gpio_npcx` feels wrong. 

`npcx_miwu_interrupt_configure` API still supports `NPCX_MIWU_MODE_DISABLED` as `mode` argument.

As @pabigot suggested such changes, if justified, should be put in a separate commit.

MulinChao replied:
Thanks for the comments. I will remove NPCX_MIWU_MODE_DISABLED and split it into two CLs.

MulinChao replied:
Done.

mnkp commented on drivers/interrupt_controller/intc_miwu.c:
```diff
@@ -333,8 +328,8 @@ int npcx_miwu_manage_dev_callback(struct miwu_dev_callback *cb, bool set)
 		for (i = 0; i < NPCX_MIWU_GROUP_COUNT; i++) {                  \
-			NPCX_WKPCL(base, i) = 0xFF;                            \
 			NPCX_WKEN(base, i) = 0;                                \
+			NPCX_WKPCL(base, i) = 0xFF;                            \
```
Again, I'm not familiar with Nuvoton Platform, however most SoCs will trigger interrupt immediately when it is enabled if the status register was dirty. That's why it's common to clear interrupt status register shortly before it's enabled. Changing the sequence here feels wrong.

keith-zephyr replied:
From the Nuvoton datasheet, bits set in the WKEN register enable the interrupt request, bits clear disable the interrupt request.  So this code is disabling all interrupts, then clearing any stale interrupts.  I believe the sequencing here is correct.

MulinChao replied:
As @keith-zephyr mentioned, the sequenced is disabling all interrupts, then clearing any stale interrupts for sure. If there is a level type interrupt issued, the original source cannot clear the pending bit successfully since it was trigger immediately.

MulinChao committed:
driver: gpio: npcx: simpler approach for GPIO_INT_MODE_DISABLED.

This CL uses a simpler configuration approach that turns GPIO's
interrupts off instead of calling npcx_miwu_interrupt_configure
with NPCX_MIWU_MODE_DISABLED.

Signed-off-by: Mulin Chao <mlchao@nuvoton.com>

MulinChao committed:
driver: intc: npcx: remove 'interrupt disable' in miwu configure func.

In this CL, npcx_miwu_interrupt_configure is no longer responsible for
turning the interrupt off. Although the default state of WK_EN is
disabled, the users still have the chance to configure them when WK_EN
is enabled via npcx_miwu_irq_enable(). Hence, this CL also ensures that
WK_EN is disabled before configuring them.

Signed-off-by: Mulin Chao <mlchao@nuvoton.com>

pabigot commented on drivers/interrupt_controller/intc_miwu.c:
```diff
@@ -186,60 +183,60 @@ int npcx_miwu_interrupt_configure(const struct npcx_wui *wui,
-		/* Clear MIWU enable bit */
-		NPCX_WKEN(base, wui->group) &= ~pmask;
+	/* If WK_EN is enabled, turn it off before configuring */
+	if (npcx_miwu_irq_get_state(wui)) {
```
Is it worth making this conditional?  That is, why not just call `npcx_miwu_irq_disable(wui)`?

If it was enabled on input, should it be re-enabled before leaving?  Just in case somebody's relying on the previous behavior.

MulinChao replied:
> Is it worth making this conditional? That is, why not just call `npcx_miwu_irq_disable(wui)`?
> 
Although I have checked all cases meet the condition that WK_EN is zero before configuration, I still think it's worth adding it since the symptom is very hard to debug. (It might issue an interrupt unexpectedly.) And yes, I will call npcx_miwu_irq_disable(wui) directly instead of it.

> If it was enabled on input, should it be re-enabled before leaving? Just in case somebody's relying on the previous behavior.
Re-enable gpio or the other miwu input source before leaving this function might cause problems. For example,  the power rails of gpio's PU or PD might be gone (ie. floating) during booting. At this moment, ec has a chance to jump into ISR and calls some routines which are not initialized properly. (On Chromebook, gpio initializations have the highest priority than the other devices during booting.)

I think the timing to turn on miwu's interrupt which ISR relies on the other device is not this driver's responsibility. We should let the other device driver or upper layer decide when to enable it via npcx_miwu_enable(wui).

MulinChao replied:
Done.

pabigot reviewed (commented):
Looks better, but I'm still not sure it's right.  Please confirm this does what you intend.
